test/CachedQuery.test.ts

```diff
@@ -1,311 +1,235 @@
         const result2 = await query.run(1).run();
-        expect(fetchFunction.mock.calls.length).toBe(2);
-        expect(result2.isLeft()).toBe(true);
-
-        expect(
-          result2.swap().getOrElse('') !== result.swap().getOrElse('')
-        ).toBe(true);
+        expect(result2).toEqual(left('nope2'));
+        fetch.assertExhausted();
       });
     });
   });
 
   describe('available', () => {
     describe('run()', () => {
       it('viene chiamata una volta la fetch e run torna il risultato successful', async () => {
-        const fetch = new StatefulFetch({
-          order: 'alwaysSuccess',
-          resultType: 'alwaysTheSame'
-        }).fetch;
-        const { fetchFunction, result } = await runQueryShallow(
-          fetch,
-          1,
-          available
-        );
-        expect(fetchFunction.mock.calls.length).toBe(1);
-        expect(result.isRight()).toBe(true);
+        const fetch = makeTestFetch([(n: number) => taskEither.of(n * 2)]);
+        const query = queryShallow(fetch, available);
+        const result = await query.run(1).run();
+        expect(result).toEqual(right(2));
+        fetch.assertExhausted();
       });
 
       it('viene chiamata una volta una fetch che fallisce sempre e run ritorna il fallimento', async () => {
-        const fetch = new StatefulFetch({
-          order: 'alwaysFailure',
-          resultType: 'alwaysTheSame'
-        }).fetch;
-        const { fetchFunction, result } = await runQueryShallow(
-          fetch,
-          1,
-          available
-        );
-        expect(fetchFunction.mock.calls.length).toBe(1);
-        expect(result.isLeft()).toBe(true);
+        const fetch = makeTestFetch([(_: number) => fromLeft('nope')]);
+        const query = queryShallow(fetch, available);
+        const result = await query.run(1).run();
+        expect(result).toEqual(left('nope'));
+        fetch.assertExhausted();
       });
 
       it('dopo avere eseguito run con successo una volta, viene chiamata di nuovo run, la fetch NON esegue e viene ritornato il precedente risultato', async () => {
-        const fetch = new StatefulFetch({
-          order: 'alwaysSuccess',
-          resultType: 'alwaysDifferent'
-        }).fetch;
-        const { fetchFunction, result, query } = await runQueryShallow(
-          fetch,
-          1,
-          available
-        );
-        expect(fetchFunction.mock.calls.length).toBe(1);
-        expect(result.isRight()).toBe(true);
+        const fetch = makeTestFetch([(n: number) => taskEither.of(n * 2)]);
+        const query = queryShallow(fetch, available);
+        const result = await query.run(1).run();
+        expect(result).toEqual(right(2));
         const result2 = await query.run(1).run();
-        expect(fetchFunction.mock.calls.length).toBe(1);
-        expect(result2.getOrElse('') === result.getOrElse('')).toBe(true);
+        expect(result2).toEqual(right(2));
+        fetch.assertExhausted();
```

Comment:
In this case also `expect(result2).toEqual(right(2))` seems like not exhaustive, what is telling us that the result is not coming from the same function being called `n` times? Probably we should give it the chance to fail by giving it an array of 2 or more fetch results and checking that `itIsNotExhausted` and that the result is the first one

Reply:
> same function being called n times

impossible given a correct implementation of `makeTestFetch`

Reply:
but given said implementation it is also impossible for this test to fail, am I wrong?

Reply:
is this basically the same question as above? `assertExhausted` implies `result2` equality and the other way around? If yes, I answered above, if not I'm not understanding

Reply:
it is in part, it is more linked to the last comment I made, my concern is that if we want to keep the current implementation where the `fetch` function cannot be called a number of times > `results.lenght` then our test should verify that the query `doesNotThrow`. Alternatively, we could change the util to return a `left(out of range)` to test against

Reply:
anyway it's a "questione di lana caprina"

Reply:
> our test should verify that the query doesNotThrow

ah ok got it now: I already verified that if that error is thrown, the test fails. So we should be covered